### ISSUE TITLE
Read-only Queries which return partial results if some components are unitialized

### ISSUE THREAD

**Issue by ElliottjPierce** [bevyengine/bevy#16923]
# TLDR

Current queries, even readonly queries, need mutable access to the world, and the existing, truly immutable, EntityRef api is slow and cumbersome in comparison. I am not aware of anything conceptual preventing truly immutable queries. Filling this gap would improve performance and readability of readonly world access, improving bevy's inter-world communication and possibly enabling performance improvements in the render world and other places.

## What problem does this solve or what need does it fill?

Currently, every way to query a world requires mutable access to the world. This is because queries need to be able to register components to form their access lists. This enables parallelization and can lead to other performance benefits by narrowing down archetypes etc. However, it does not make semantic sense that a read-only query should need write access to a world. 

For example, if a query needs a &Transform, and the Transform component has not been registered, the whole query could be skipped. Similar logic can be applied to Option<&Transform>, With<Transform>, Without<Transform>, etc. In general, having a unregistered component either makes the query empty, makes a filter irrelevant, or forces an Option to be None. It should be possible to implement true, read-only queries, even if we loose some performance from being able to skip some archetypes, etc.

Additionally, having true, read-only queries could enable further parallelization when communicating between worlds. In my use case, I have a model world and a view world, where the view should be able to query the model without needing mutable access. Also, in bevy's sub-apps, especially the render world, this may enable more parallel reads from the main world, though honestly I don't know the render world very well. 

## What solution would you like?

I don't know what would be best, but here are some options:

- give bevy::ecs::query::ReadOnlyQueryData a function for getting its fetch directly and fully immutably.
- create a new fully readonly query trait ecosystem built on EntityRef
- use some kind of unsafe magic to synchronize the existing component registration in the current query traits.
- use some synchronizer to make World::register_component need only immutable access to the world

## What alternative(s) have you considered?

EntityRef exists to fulfill this exact problem; however, it is cumbersome compared to the elegance of creating a query. Additionally, World::iter_entities can not be sped up by skipping some Archetypes, etc. While not always possible, a readonly query abstraction would sometimes be able to skip archetypes and could be much much faster.

## Additional context

My use case is a multiplayer framework that I'm making as a learning tool. I want the server and client sides to be physically separated  by putting them in different worlds. This lets each client world focus only on one player and will hopefully make split-screen, listen-servers, and the like much easier. When the clients worlds look at the server world (which could be replicated remotely), the clients should be able to query the server world for information. While this is possible with the existing EntityRef api, queries are much easier to work with. Let me know if there is a better way to do this or if I am missing something.

**Comment by hymm** [comment]
`World::try_query` was recently added and will come in 0.16. https://dev-docs.bevyengine.org/bevy/ecs/world/struct.World.html#method.try_query. Does that fulfill your requirements or do you need something else?

**Comment by ElliottjPierce** [comment]
That looks awesome and is very nearly what I'm looking for. It looks like when try_query returns None, it is because some components were not registered but were referenced in the query. Instead, I was hoping for a silent fail that still gave as much of the query results as could be processed.

I'm hoping for an interface that has the following behavior when MyComponent is referenced in the query but not registered:
- For WorldQueries like &MyComponent, With<MyComponent>, Ref<MyComponent>, etc: the query is simply empty.
- For Has<MyComponent>, Option<&MyComponent>, etc: the default is given (false, None).
- For Without<MyComponent> or similar: the query ignores the missing registration since all entities are without MyComponent.

For example, `Query<(&Transform, &MyComponent)>` would work but be empty, `Query(&Transform, Option<&MyComponent)>` would just have the option be None for every entity, and `Query<&Transform, Without<MyComponent>>` would be the same as `Query<&Transform>`.

The queries would need to update its state when components are registered, but I think that would be possible. Any feedback is appreciated.

**Comment by ElliottjPierce** [comment]
Closing in favor of #18155.
